File: frontend/src/schemas/store.ts
import { z } from 'zod'

import { parseMoneyToCents } from '@/utils/money'

export const StoreSchema = z.object({
  id: z.number(),
  display_name: z.string(),
  slug: z.string(),
  created_at: z.string(),
  listing_count: z.number(),
})

export type Store = z.infer<typeof StoreSchema>

export const StoresSchema = z.array(StoreSchema)

/** The public storefront: no owner, no ids a shopper has no use for. */
export const StorefrontSchema = z.object({
  display_name: z.string(),
  slug: z.string(),
  created_at: z.string(),
  listing_count: z.number(),
})

export type Storefront = z.infer<typeof StorefrontSchema>

export const StoreFormSchema = z.object({
  display_name: z
    .string()
    .trim()
    .min(2, 'Store name must be at least 2 characters')
    .max(120, 'Store name is too long'),
})

export type StoreFormValues = z.infer<typeof StoreFormSchema>

export const UploadPolicySchema = z.object({
  configured: z.boolean(),
  max_bytes: z.number(),
  max_images: z.number(),
  allowed_types: z.array(z.string()),
})

export type UploadPolicy = z.infer<typeof UploadPolicySchema>

export const UploadTicketSchema = z.object({
  upload_url: z.string(),
  key: z.string(),
  public_url: z.string(),
  expires_in: z.number(),
})

export type UploadTicket = z.infer<typeof UploadTicketSchema>

export const ListingSchema = z.object({
  id: z.number(),
  store_id: z.number(),
  slug: z.string(),
  title: z.string(),
  brand: z.string(),
  image: z.string().nullable(),
  price_cents: z.number(),
  list_price_cents: z.number().nullable(),
  stock: z.number(),
  variant_count: z.number(),
  rating_avg: z.number(),
  rating_count: z.number(),
  created_at: z.string(),
})

export type Listing = z.infer<typeof ListingSchema>

export const ListingPageSchema = z.object({
  items: z.array(ListingSchema),
  total: z.number(),
  page: z.number(),
  page_size: z.number(),
})

export type ListingPage = z.infer<typeof ListingPageSchema>

const money = (message: string) =>
  z.string().refine((value) => parseMoneyToCents(value) !== null, message)

/** What the form holds, which is not what the API takes: prices are typed as dollars and
 *  the option axis is named once for the whole listing rather than repeated per row. */
export const ListingFormSchema = z
  .object({
    title: z.string().trim().min(3, 'Give the listing a title').max(300, 'Title is too long'),
    brand: z.string().trim().min(1, 'Who makes it?').max(120, 'Brand name is too long'),
    category_slug: z.string().min(1, 'Pick a category'),
    description: z.string().max(4000, 'Description is too long'),
    bullets: z.array(z.object({ text: z.string().max(200, 'Keep bullets under 200 characters') })),
    option_name: z.string().max(60, 'Option name is too long'),
    variants: z
      .array(
        z.object({
          option_value: z.string().max(60, 'Option value is too long'),
          price: money('Enter a price like 24.99'),
          list_price: z.union([money('Enter a price like 29.99'), z.literal('')]),
          stock: z.coerce
            .number({ message: 'Enter a whole number' })
            .int('Enter a whole number')
            .min(0, 'Stock cannot be negative')
            .max(100_000, 'That is a lot of stock'),
        }),
      )
      .min(1)
      .max(8, 'Up to 8 options per listing'),
    images: z
      .array(z.object({ key: z.string(), url: z.string() }))
      .min(1, 'Add at least one photo')
      .max(6, 'Up to 6 photos per listing'),
  })
  .superRefine((values, context) => {
    values.variants.forEach((variant, index) => {
      const price = parseMoneyToCents(variant.price)
      if (price === 0) {
        context.addIssue({
          code: 'custom',
          path: ['variants', index, 'price'],
          message: 'Price must be more than zero',
        })
      }
      const listPrice = variant.list_price ? parseMoneyToCents(variant.list_price) : null
      if (price !== null && listPrice !== null && listPrice <= price) {
        context.addIssue({
          code: 'custom',
          path: ['variants', index, 'list_price'],
          message: 'The was-price has to be higher than the price',
        })
      }
    })

    if (values.variants.length < 2) return

    // More than one option means each needs a label, or the buy box shows identical rows.
    if (!values.option_name.trim()) {
      context.addIssue({
        code: 'custom',
        path: ['option_name'],
        message: 'Name the option so shoppers can tell the choices apart',
      })
    }
    const seen = new Set<string>()
    values.variants.forEach((variant, index) => {
      const value = variant.option_value.trim()
      if (!value) {
        context.addIssue({
          code: 'custom',
          path: ['variants', index, 'option_value'],
          message: 'Give this option a value',
        })
        return
      }
      if (seen.has(value)) {
        context.addIssue({
          code: 'custom',
          path: ['variants', index, 'option_value'],
          message: 'Two options cannot have the same value',
        })
      }
      seen.add(value)
    })
  })

export type ListingFormValues = z.input<typeof ListingFormSchema>

File: backend/app/schemas/store.py
from datetime import datetime

from pydantic import BaseModel, Field, field_validator, model_validator

from app.services.constants import (
    ALLOWED_IMAGE_TYPES,
    MAX_IMAGES_PER_LISTING,
    MAX_UPLOAD_BYTES,
    MAX_VARIANTS_PER_LISTING,
)

# A listing priced above this is almost certainly a units mistake (dollars typed into a
# cents field). Rejecting it at the boundary beats a $4,999,900.00 tile in the grid.
MAX_PRICE_CENTS = 5_000_000
MAX_STOCK = 100_000


class StoreCreate(BaseModel):
    display_name: str = Field(min_length=2, max_length=120)

    @field_validator("display_name")
    @classmethod
    def _trim(cls, value: str) -> str:
        trimmed = value.strip()
        if len(trimmed) < 2:
            raise ValueError("Store name must be at least 2 characters")
        return trimmed


class StoreOut(BaseModel):
    id: int
    display_name: str
    slug: str
    created_at: datetime
    listing_count: int = 0


class StorefrontOut(BaseModel):
    """The public face of a store. No user id, no counts a shopper cannot already see."""

    display_name: str
    slug: str
    created_at: datetime
    listing_count: int


class UploadPolicy(BaseModel):
    """Lets the sell screen render the cap, the accepted types and the not-configured
    notice without hardcoding any of them in the client."""

    configured: bool
    max_bytes: int = MAX_UPLOAD_BYTES
    max_images: int = MAX_IMAGES_PER_LISTING
    allowed_types: list[str] = sorted(ALLOWED_IMAGE_TYPES)


class UploadTicketIn(BaseModel):
    # Images are stored under the store's own prefix, so the ticket has to say which one.
    store_id: int
    content_type: str
    size_bytes: int = Field(gt=0, le=MAX_UPLOAD_BYTES)

    @field_validator("content_type")
    @classmethod
    def _allowed(cls, value: str) -> str:
        if value not in ALLOWED_IMAGE_TYPES:
            raise ValueError(f"Images must be one of: {', '.join(sorted(ALLOWED_IMAGE_TYPES))}")
        return value


class UploadTicketOut(BaseModel):
    upload_url: str
    key: str
    # Where the image will be readable once the PUT succeeds, so the form can preview the
    # real URL rather than a local blob it would have to revoke later.
    public_url: str
    expires_in: int


class VariantIn(BaseModel):
    """One purchasable option. Price and stock live here, never on the listing."""

    attrs: dict[str, str] = Field(default_factory=dict)
    price_cents: int = Field(gt=0, le=MAX_PRICE_CENTS)
    list_price_cents: int | None = Field(default=None, gt=0, le=MAX_PRICE_CENTS)
    stock: int = Field(ge=0, le=MAX_STOCK)

    @model_validator(mode="after")
    def _list_price_is_a_discount(self):
        if self.list_price_cents is not None and self.list_price_cents <= self.price_cents:
            raise ValueError("Was-price must be higher than the selling price")
        return self


class ListingCreate(BaseModel):
    title: str = Field(min_length=3, max_length=300)
    brand: str = Field(min_length=1, max_length=120)
    category_slug: str = Field(min_length=1, max_length=140)
    description: str = Field(default="", max_length=4000)
    bullets: list[str] = Field(default_factory=list, max_length=6)
    image_keys: list[str] = Field(min_length=1, max_length=MAX_IMAGES_PER_LISTING)
    variants: list[VariantIn] = Field(min_length=1, max_length=MAX_VARIANTS_PER_LISTING)

    @field_validator("bullets")
    @classmethod
    def _drop_blanks(cls, value: list[str]) -> list[str]:
        return [bullet.strip()[:200] for bullet in value if bullet.strip()]

    @model_validator(mode="after")
    def _distinct(self):
        if len(set(self.image_keys)) != len(self.image_keys):
            raise ValueError("The same image was attached twice")
        # Two variants with identical options would render as duplicate rows in the buy
        # box with no way to tell them apart.
        options = [tuple(sorted(variant.attrs.items())) for variant in self.variants]
        if len(set(options)) != len(options):
            raise ValueError("Two options have the same values")
        return self


class ListingOut(BaseModel):
    id: int
    store_id: int
    slug: str
    title: str
    brand: str
    image: str | None
    price_cents: int
    list_price_cents: int | None
    stock: int
    variant_count: int
    rating_avg: float
    rating_count: int
    created_at: datetime
